### PR TITLE
Fix import errors in Python3.5.0 and Python3.5.1

### PR DESCRIPTION
**Pull request by sile**
`typing` module of Python3.5.0 and Python3.5.1 has not defined some classes that are provided in other Python versions (e.g., `typing.Type` class).
Because of this, in Python3.5.1 (or 3.5.0), `import optuna` fails as follows
```console
$ docker run --rm -it python:3.5.1 /bin/bash
# git clone -b master git://github.com/pfnet/optuna
# cd optuna

$ pip install --upgrade pip
# pip install .
# python -c 'import optuna; print("OK")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/optuna/optuna/__init__.py", line 2, in <module>
    from optuna import dashboard  # NOQA
  File "/optuna/optuna/dashboard.py", line 26, in <module>
    import optuna.study
  File "/optuna/optuna/study.py", line 17, in <module>
    from typing import Type  # NOQA
ImportError: cannot import name 'Type'
```

However, these missing classes are only needed for type checking by `mypy` and unnecessary for ordinal program executions.
In this PR, I changed to import those classes only when `optuna.types.TYPE_CHECKING` flag is set.
By this modification, `import optuna` became to work well in Python3.5.0 and Python3.5.1.
```console
$ docker run --rm -it python:3.5.1 /bin/bash
# git clone -b fix-import-typing-error git://github.com/pfnet/optuna
# cd optuna

# pip install --upgrade pip
# pip install .
# python -c 'import optuna; print("OK")'
OK
```

### CODE REVIEW

File: optuna/study.py [hunk 1]
```diff
@@ -7,22 +7,25 @@
...
-from typing import Any  # NOQA
-from typing import Callable  # NOQA
-from typing import Dict  # NOQA
-from typing import List  # NOQA
-from typing import Optional  # NOQA
-from typing import Set  # NOQA
-from typing import Tuple  # NOQA
-from typing import Type  # NOQA
-from typing import Union  # NOQA
+from typing import Callable
```
Comment (toshihikoyanase) [inline optuna/study.py]: Can we move this import statement `Callable` and the definition of `ObjectiveFuncType` to the `if types.TYPE_CHECKING` block? `Callable` is used to define `ObjectiveFuncType` at L30, and `ObjectiveFuncType` is only used for type checking. So, I think they are not required to execute code.

Comment (sile) [reply optuna/study.py]: Yes, we can. Thank you for pointing it out.

Comment (sile) [reply optuna/study.py]: Fixed: https://github.com/pfnet/optuna/pull/326/commits/ec275748e7bf404e37f5e33dea542e030d6b7ff4

**Comment by sile** [comment]
@toshihikoyanase Thank you for reviewing. I merged `master` branch and resolved the conflict.

**Pull request merged**